### PR TITLE
Improve (de)serialization and validation capabilities

### PR DESCRIPTION
**Pull request by sscherfke**
Improve (de)serialization and validation capabilities by adding hooks that can automatically update a class’ attributes and that can change the way how objects are serialized by `asdict()`.

See: #649

# Pull Request Check List

This is just a friendly reminder about the most common mistakes.  Please make sure that you tick all boxes.  But please read our [contribution guide](https://www.attrs.org/en/latest/contributing.html) at least once, it will save you unnecessary review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing left to do.

- [x] Added **tests** for changed code.
- [x] New features have been added to our [Hypothesis testing strategy](https://github.com/python-attrs/attrs/blob/master/tests/strategies.py).
- [x] Changes or additions to public APIs are reflected in our type stubs (files ending in ``.pyi``).
    - [x] ...and used in the stub test file `tests/typing_example.py`.
- [x] Updated **documentation** for changed code.
    - [x] New functions/classes have to be added to `docs/api.rst` by hand.
    - [x] Changes to the signature of `@attr.s()` have to be added by hand too.
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded).  Find the appropriate next version in our [``__init__.py``](https://github.com/python-attrs/attrs/blob/master/src/attr/__init__.py) file.
- [x] Documentation in `.rst` files is written using [semantic newlines](https://rhodesmill.org/brandon/2012/one-sentence-per-line/).
- [x] Changes (and possible deprecations) have news fragments in [`changelog.d`](https://github.com/python-attrs/attrs/blob/master/changelog.d).

If you have *any* questions to *any* of the points above, just **submit and ask**!  This checklist is here to *help* you, not to deter you from contributing!

### CODE REVIEW

File: tests/test_make.py [hunk 1]
```diff
@@ -2131,3 +2138,58 @@ def __setstate__(self, state):
...
+        assert results == [
+            ('x', int, True),
+            ('y', str, True),
+        ]
+
+    @pytest.mark.skipif(PY2, reason="Auto-detection is Python 3-only.")
+    def test_hook_applied_modify_attrib(self):
+        """The transform hook must be able to modify attributes."""
+        def hook(name, ca, type):
+            ca.converter = type
```
Comment (wsanchez) [inline tests/test_make.py:2187]: Hrm… if I were writing a hook like this, I'd be type hinting everything, and the type of `ca` here is `_CountingAttr`, which is private.  That seems like a problem.

IIRC we've avoided making `_CountingAttr` public, so this API would require us to at least define an interface, no?

Comment (sscherfke) [reply tests/test_make.py:2187]: Yes, that’s true. I also stumbled upon this when I was writing the docstring. Making an Interface seems like a reasonable idea. We wouldn’t to expose the actual `_CountingAttr` than but mypy would be happy nonetheless.

**Comment by sscherfke** [comment]
I have a lot more uncommited code, but some (important) parts of it require at least Python 3.6.

When are you planning to drop support for Python 2 (and maybe even 3.5)?  This would make a few things easier and if that’s not too far into the future, I would gladly wait a little. :)

**Comment by wsanchez** [comment]
> When are you planning to drop support for Python 2 (and maybe even 3.5)? This would make a few things easier and if that’s not too far into the future, I would gladly wait a little. :)

I suggest you file an Issue called "[RFC] Drop support for Python 2" and have that discussion there, as it's not specific to this ticket. (And perhaps another for 3.5.)

**Comment by hynek** [comment]
I'm sorry I'm offline for the next week, but there's been a feature request asking for `_CountingAttr` to become public already so this might be two birds with one stone.

As for Python 2: https://www.attrs.org/en/stable/python-2.html

As much as I don't like it, it's one of our main features over dataclasses ATM.

_Once_ we go Python 3-only, it will definitely not be 3.5. :)

**Comment by sscherfke** [comment]
What is your preferred approach for tests that run (and parse) only with python 3.6 and above?

**Comment by hynek** [comment]
https://github.com/python-attrs/attrs/blob/3e3d84216c5c7b6b451610ebbe8f5a7ed9c3209b/conftest.py#L17

**Comment by sscherfke** [comment]
This PR will only focus on the hooks. I have removed the converters and validators and will re-propose them in another PR when this one is accepted.

**Comment by sscherfke** [comment]
`_CountingAttr` (or an interface for it) need to be made public for this PR.

What would be your prefered approach?  Create a new Interface that `_CountingAttr` inherits or rename `_CountingAttr` to `CountingAttr`? Should also be imported in `__init__.py` in that case?

**Comment by wsanchez** [comment]
> `_CountingAttr` (or an interface for it) need to be made public for this PR.
> 
> What would be your prefered approach? Create a new Interface that `_CountingAttr` inherits or rename `_CountingAttr` to `CountingAttr`? Should also be imported in `__init__.py` in that case?

Again, we've resisted making `_CountingAttr` in the past because it's not an API we want to commit to long-term, so an interface that is limited to what we *are* willing to support seems best, leaving `_CountingAttr` itself as an implementation detail.

That assumes that @hynek is amenable to the resulting public interface.

**Comment by hynek** [comment]
The main issue here is that I don't want it to be "counting" anymore if not necessary (ordered dicts), cf #643. I guess we'll have to extract a sub-interface?

**Comment by sscherfke** [comment]
An Interface would suffice since it will (at least in this PR) only be used for documentation and (mypy) typing. I can try to extract an Interface.

**Comment by sscherfke** [comment]
Hm, `_CountingAttr` is so "simple" and specific, that I don’t know which attributes/meths to extract into an interface.

**Comment by wsanchez** [comment]
> Hm, `_CountingAttr` is so "simple" and specific, that I don’t know which attributes/meths to extract into an interface.

I'd start with: what exactly is the minimum needed for this PR to function?

**Comment by sscherfke** [comment]
*I* only want to add a converter to the attrib, but instances of `_CountingAttr` (I’d call the interface `MutableAttr` since it is, in contrast to `Attribute`, mutable) are passed to the hook so it would not be very long-sighted to only add the converter attribute to the interface.  I think at least `converter`, `metadata`, `validator()` and `default()` should be made public.  Changing the other attributes in the hook might not make too much sense.

**Comment by hynek** [comment]
I guess a fairer name would be `AttributeDefinition` but I wonder if we couldn't spin it in a way that the hook receives `Attributes` but is free to replace them with something else? The would be definitely cleaner and less invasive.

**Comment by sscherfke** [comment]
I’m not sure if I understand what you mean.  The whole purpose of the hook is to be invasive and let the user mess with the attribute definition.

**Comment by hynek** [comment]
I mean that due to stuff like resolving decorator-based defaults and validators we can’t really provide a good interface? My hope would be that the hook would get a list of `Attribute`s will fully resolved validators etc and would return a list of `Attributes` that may replace some of them or even remove some?

**Comment by sscherfke** [comment]
Thanks for the clarification. I will think about this and see if it works.

**Comment by hynek** [comment]
Yeah, please let me know what you think. The upsides are mainly that we don’t have to introduce a new “in-between type” that would have to be mutated by the users. That smells like bugs and edge cases to me. I’m open to add more helper APIs to `Attribute`, though!

**Comment by sscherfke** [comment]
This was an excelent idea of you, @hynek.  :) The hook is now purely functional. It takes an `Attribute` and returns an `Attribute`.  The only thing that we’d need to change now is to make `Attribute._assoc` public in order to allow people to evolve an attribute.

**Comment by sscherfke** [comment]
ping :)

**Comment by sscherfke** [comment]
ping :)

**Comment by hynek** [comment]
Yeah I'm sorry, I really had to focus on landing all the `__setattr__` stuff (#663 still remains) because I was procrastinating on this stuff for months and it was the last hurdle before starting to work on "import attrs", because it affects default values. Once that lands, I urgently need to get 20.1 out of the door this, the `__eq__` PR, and "import attrs" will be my next priorities for 20.2 which I hope to get out of the door _much_ faster.

I was basically running behind a carrot all year and had to draw a line now. Sorry it affected you. :(((

**Comment by sscherfke** [comment]
No worries.  I know this problem.  Just wanted to make sure I’m not forgotten. :)

**Comment by hynek** [comment]
No worries, I **really** want to open attrs to your use cases!

**Comment by hynek** [comment]
(you may want to start rebasing tho, sadly the innards have quite changed :()

**Comment by sscherfke** [comment]
I think the PR is now complete.

When this gets merge, I will create another one with some new validators, a default transform_hook, and a new section in the *Examples* documentation.

**Comment by hynek** [comment]
At a first glance this look reasonable!

Just to double check: have you been able to implement your PoCs with this kind of code? Because as it stands here, there is no real justification for it so I just wanna make sure you’re 100% positive this is what you need and it’s sufficient. :)

**Comment by sscherfke** [comment]
Yes, I already tested the hooks with an API client (JSON (de)serialization to/from attrs classes) and the same hook should work for my typed-settings PoC as well.

Instead of creating a second PR, I can add its contents directly to this one. This might make the benefits of this PR more clear.

**Comment by sscherfke** [comment]
I’ve added a draft for additional validators and (auto) converters. This would allow us to use attrs directly for

- API clients (load attrs classes from json and serialize request data to json)
- Typed settings loader

Here is an example:

```python
from datetime import datetime
from enum import Enum
from typing import List, Set
import json

from attr.converters import auto_convert_hook
from attr.validators import le, maxlen
import attr


class LeEnum(Enum):
    spam = 'Le spam'
    eggs = 'Le eggs'


@attr.frozen(transform_hook=auto_convert_hook)
class Child:
    x: int = attr.ib()
    y: int = attr.ib(converter=float)


@attr.frozen(kw_only=True, transform_hook=auto_convert_hook)
class Parent:
    child: Child
    a: float
    b: float = attr.field(default=3.14, validator=le(2))
    c: LeEnum
    d: datetime
    e: 'List[Child]'
    f: Set[datetime]


DATA = {
    'a': '1',
    'b': '2',
    'c': 'Le spam',
    'd': '2020-05-04T13:37:00',
    'e': [
        {
            'x': '23',
            'y': '42',
        },
    ],
    'f': ['2020-05-04T13:37:00', '2020-05-04T13:37:00'],
    'child': {
        'x': '23',
        'y': '42',
    },
}
p = Parent(**DATA)
print(p)
d = attr.asdict(p, hook=attr.converters.serialize_hook)
print(d)

jd = json.dumps(d)
print(Parent(**json.loads(json.dumps(d))) == p)
```

The validators and converters should cover the most common use cases. I’d like to make the `auto_convert_hook()` more customizable, though, so that users could add additional types without having to create a completely new hook.

**Comment by sscherfke** [comment]
There is no a `make_auto_converter()` factory that allows you to extend the bundled `auto_convert()` hook. 
I also added a few tests for this.

What do you think of the general direction this PR is taking? Shall I continue with adding tests and docs?

**Comment by si14** [comment]
A note from a silent observer: this PR looks like a fabulous addition to `attrs` and I'm really looking forward to using it. Thanks for your great work, @sscherfke and @hynek !

File: src/attr/_make.py [hunk 2]
```diff
@@ -484,6 +487,9 @@ def _transform_attrs(cls, these, auto_attribs, kw_only, collect_by_mro):
         for attr_name, ca in ca_list
     ]
 
+    if transform_hook is not None:
+        own_attrs = [transform_hook(cls, a) for a in own_attrs]
```
Comment (hynek) [inline src/attr/_make.py:491]: Out of curiosity: why did you choose to call them one by one and not all at once?

I have a gut feeling that an `own_atts = transform_hooks(cls, own_attrs)` wouldn't be just faster (less method calls) but also allow for more use-cases where you need see all attributes? 🤔

File: changelog.d/653.change.rst [hunk 3]
```diff
@@ -0,0 +1,5 @@
+``attr.s()`` now has a *transform_hook* that is called for each
```
Comment (hynek) [inline changelog.d/653.change.rst:1]: We use [semantic newlines](http://rhodesmill.org/brandon/2012/one-sentence-per-line/). :)

File: docs/api.rst [hunk 4]
```diff
@@ -24,7 +24,7 @@ Core
 
 .. autodata:: attr.NOTHING
 
-.. autofunction:: attr.s(these=None, repr_ns=None, repr=None, cmp=None, hash=None, init=None, slots=False, frozen=False, weakref_slot=True, str=False, auto_attribs=False, kw_only=False, cache_hash=False, auto_exc=False, eq=None, order=None, auto_detect=False, collect_by_mro=False, getstate_setstate=None, on_setattr=None)
+.. autofunction:: attr.s(these=None, repr_ns=None, repr=None, cmp=None, hash=None, init=None, slots=False, frozen=False, weakref_slot=True, str=False, auto_attribs=False, kw_only=False, cache_hash=False, auto_exc=False, eq=None, order=None, auto_detect=False, collect_by_mro=False, getstate_setstate=None, on_setattr=None, transform_hook=None)
```
Comment (hynek) [inline docs/api.rst:27]: Do you think we could come up with a better name for the hook? While through the lens of this PR it makes complete sense, I feel its a bit generic next to the other options. It also bleeds out the name of the function that calls it which doesn't make much sense to the users. Would you agree?

File: src/attr/_funcs.py [hunk 5]
```diff
@@ -13,6 +13,7 @@ def asdict(
     filter=None,
     dict_factory=dict,
     retain_collection_types=False,
+    hook=None,
```
Comment (hynek) [inline src/attr/_funcs.py:16]: Same here, I'm afraid we need long arg names. :)

File: src/attr/converters.py [hunk 6]
```diff
@@ -83,3 +86,117 @@ def default_if_none_converter(val):
             return default
 
     return default_if_none_converter
+
+
+def convert_attrs(cls):
```
Comment (hynek) [inline src/attr/converters.py:91]: This is not our style of writing docstrings. 😇

(Check out https://github.com/python-attrs/attrs/blob/master/.github/CONTRIBUTING.rst#code – I know your style is the CPython one but I can't stand it because the rationale per PEP 257 is that Guido's Emacs can't format paragraphs otherwise 🤪).

**Review by hynek — CHANGES_REQUESTED** [review]
So overall I'm quite happy how it's turning out!

The following needs to be done:

- I think we need a better name for `transform_hook`, I explain it better in the comments.
- Please remove the converters/validators to keep this PR small. Your actual feature is literally a two line change but due to its impact on the core of attrs it ripples throughout the whole code base. 🙈 We'll do it in a separate PR. At this point I'd like to get it in and build on it.
- You can start adding basic docs, yes! Maybe a short chapter on its own? We'll have to edit/rewrite it probably but we gotta start somewhere and you're the best person for it since the whole thing has been started by your own motivations. :)
- I need to get our a bug release ASAP, but this PR is on the top of my big-things-for-attrs-queue.

I'm very excited by all of this!

File: src/attr/_make.py (hunk 2 above)
Comment (sscherfke) [reply src/attr/_make.py:491]: This is a good point. When I started work on this feature, I called the hook in another place where I *needed* to call it once for each attrib. This is now no longer necessary and my default auto-convert-hook would actually benefit from making this one-call-for-all. :+1:

File: src/attr/converters.py (hunk 6 above)
Comment (sscherfke) [reply src/attr/converters.py:91]: Alright, I will fix this. :)

File: docs/api.rst (hunk 4 above)
Comment (sscherfke) [reply docs/api.rst:27]: Naming things is so hard! 🤪 Is something like `attrib_converter()` better?

File: src/attr/_funcs.py (hunk 5 above)
Comment (sscherfke) [reply src/attr/_funcs.py:16]: `value_serializer()`?

**Comment by sscherfke** [comment]
Ok, I'll rip out the validators and converters again and put them in a separate PR.

Would a new section in the `examples.rst` be a gould place for the docs or shall I create a new document?

File: src/attr/_make.py (hunk 2 above)
Comment (hynek) [reply src/attr/_make.py:491]: What was the use-case for needing to call them once? Can't you simulate it using an intermediate function?

File: docs/api.rst (hunk 4 above)
Comment (hynek) [reply docs/api.rst:27]: So this is awkward.

The reason why new-style APIs use `field` is because it's quite distinct to the soup of attrs, attribs, attr and whatnot.

Given we're marching towards a brighter future…maybe `field_transformer`? 😇

Summoning the creative juices of @wsanchez @glyph @euresti @Tinche etc 😅

File: src/attr/_funcs.py (hunk 5 above)
Comment (hynek) [reply src/attr/_funcs.py:16]: Yeah, better I think.

**Comment by hynek** [comment]
I think this is an advanced feature so either `extending.rst` or something new entirely is apt?

File: docs/api.rst (hunk 4 above)
Comment (ambv) [reply docs/api.rst:27]: +1 to `field_transformer`

File: src/attr/_make.py (hunk 2 above)
Comment (sscherfke) [reply src/attr/_make.py:491]: There never was a use case. It was just the natural way of doing it were I originally implemented it. I think there is no drawback in making it `hook(cls: type, fields: List[Attribute]) -> List[Attribute])`

Comment (hynek) [reply src/attr/_make.py:491]: Wonderful, pls do. Let’s provide the users with as much context as necessary.

**Comment by sscherfke** [comment]
I renamed the hooks, (hopefully) fixed the docs format, added more documentation, removed the validators and converters, rebased on master, and updated *versionadded* labels to 20.3.0.

**Comment by sscherfke** [comment]
Did some more polishing and fixed some formatting issues.

File: src/attr/_make.py [hunk 7]
```diff
@@ -484,6 +487,9 @@ def _transform_attrs(cls, these, auto_attribs, kw_only, collect_by_mro):
         for attr_name, ca in ca_list
     ]
 
+    if field_transformer is not None:
```
Comment (hynek) [inline src/attr/_make.py:490]: One quick (but important) thing: why are you calling this so early? I would put it at the very end. There's no reason to ignore inherited attributes, is there?

BTW another reason why calling it on all attrs is much better: the transformer has to power to add or remove attributes if it wants to! 🎉

Comment (sscherfke) [reply src/attr/_make.py:490]: Do you mean like this`
```python
    if field_transformer is not None:
        attrs = field_transformer(cls, attrs)
    return _Attributes((attrs, base_attrs, base_attr_map))
```

Comment (sscherfke) [reply src/attr/_make.py:490]: > BTW another reason why calling it on all attrs is much better: the transformer has to power to add or remove attributes if it wants to! 🎉

The docs contain an example where I use this hook to remove all float fields from a class – because, who likes floats? 🤪

Comment (sscherfke) [reply src/attr/_make.py]: Seems to work. I also added a test for this.

**Comment by sscherfke** [comment]
ping :cat2:

**Comment by hynek** [comment]
> ping 🐈

Current status:

current status

**Comment by sscherfke** [comment]
Oh, I thought you’re already back. Then have a lot of fun and I’ll be patient! 🏄

File: docs/extending.rst [hunk 8]
```diff
@@ -159,3 +159,107 @@ Here are some tips for effective use of metadata:
...
+You do this by passing a *field_transformer* hook to :func:`attr.define()` (and its friends).
+Its main purpose is to automatically add converters to attributes based on their type to aid the development of API clients and other typed data loaders.
+
+| This hook must have the signature:
+| :code:`your_hook(cls: type, fields: List[Attribute]) -> List[Attribute]`
+
+- *cls* is your class right *before* it is being converted into an attrs class.
+  This means it does not yet have the ``__attrs_attrs__`` attribute.
+
+- *fields* is a list of all :class:`attr.Attribute` instances that will later be set to ``__attrs_attrs__``.  You can modify these attributes any way you want:
```
Comment (hynek) [inline docs/extending.rst:177]:
```suggestion
- *fields* is a list of all :class:`attr.Attribute` instances that will later be set to ``__attrs_attrs__``. 
  You can modify these attributes any way you want:
```

File: docs/extending.rst [hunk 9]
```diff
@@ -159,3 +159,107 @@ Here are some tips for effective use of metadata:
...
+   ...
+   >>> from_json = {"a": 3, "b": "spam", "c": "2020-05-04T13:37:00"}
+   >>> Data(**from_json)  # ****
+   Data(a=3, b='spam', c=datetime.datetime(2020, 5, 4, 13, 37))
+
+
+Change value serialization in ``asdict()``
+------------------------------------------
+
+Attrs allows you to serialize instances of attrs classes to dicts using the :func:`attr.asdict()` function.
```
Comment (hynek) [inline docs/extending.rst:229]:
```suggestion
``attrs`` allows you to serialize instances of ``attrs`` classes to dicts using the `attr.asdict` function.
```

File: docs/extending.rst [hunk 10]
```diff
@@ -159,3 +159,107 @@ Here are some tips for effective use of metadata:
...
+   ...
+   >>> data = attr.asdict(Data(datetime.datetime(2020, 5, 4, 13, 37)))
+   >>> data
+   {'dt': datetime.datetime(2020, 5, 4, 13, 37)}
+   >>> json.dumps(data)
+   Traceback (most recent call last):
+     ...
+   TypeError: Object of type datetime is not JSON serializable
+
+To help you with this, :attr:`attr.asdict()` allows you to pass a *value_serializer* hook.  It has the signature :code:`your_hook(inst: type, field: Attribute, value: Any) -> Any`:
```
Comment (hynek) [inline docs/extending.rst:249]:
```suggestion
To help you with this, `attr.asdict` allows you to pass a *value_serializer* hook.
It has the signature

.. function:: your_hook(inst: type, field: attr.Attribute, value: typing.Any) -> typing.Any
   :noindex:
```

File: docs/extending.rst [hunk 11]
```diff
@@ -159,3 +159,107 @@ Here are some tips for effective use of metadata:
...
+   ...     b: str
+   ...     c: datetime
+   ...
+   >>> from_json = {"a": 3, "b": "spam", "c": "2020-05-04T13:37:00"}
+   >>> Data(**from_json)  # ****
+   Data(a=3, b='spam', c=datetime.datetime(2020, 5, 4, 13, 37))
+
+
+Change value serialization in ``asdict()``
+------------------------------------------
```
Comment (hynek) [inline docs/extending.rst:227]: I think Customize is better here:

```suggestion
Customize Value Serialization in ``asdict()``
---------------------------------------------
```

File: src/attr/_funcs.py [hunk 12]
```diff
@@ -32,6 +33,9 @@ def asdict(
     :param bool retain_collection_types: Do not convert to ``list`` when
         encountering an attribute whose type is ``tuple`` or ``set``.  Only
         meaningful if ``recurse`` is ``True``.
+    :param callable value_serializer: A hook that is called for every attribute
+        or dict key/value.  It receives the current instance, field and value
+        and must return the (updated) value.
```
Comment (hynek) [inline src/attr/_funcs.py:38]:
```suggestion
    :param Optional[callable] value_serializer: A hook that is called for every
        attribute or dict key/value.  It receives the current instance, field and
        value and must return the (updated) value.
```

(I'm eyeballing the line length here, I hope it works out 🤪)

File: src/attr/_funcs.py [hunk 13]
```diff
@@ -40,24 +44,36 @@ def asdict(
...
     ..  versionadded:: 16.1.0 *retain_collection_types*
+    ..  versionadded:: 20.3.0 *value_serializer*
     """
     attrs = fields(inst.__class__)
     rv = dict_factory()
     for a in attrs:
         v = getattr(inst, a.name)
+        if value_serializer is not None:
+            v = value_serializer(inst, a, v)
         if filter is not None and not filter(a, v):
```
Comment (hynek) [inline src/attr/_funcs.py:55]: Wouldn't it be better, if we ran the filters first to avoid unnecessary serialization? The order should be documented in any case.

File: src/attr/_make.py [hunk 14]
```diff
@@ -1196,6 +1208,13 @@ def attrs(
         If a list of callables is passed, they're automatically wrapped in an
         `attr.setters.pipe`.
 
+    :param field_transformer:
```
Comment (hynek) [inline src/attr/_make.py:1211]:
```suggestion
    :param Optional[callable] field_transformer:
```

File: src/attr/_make.py [hunk 15]
```diff
@@ -1196,6 +1208,13 @@ def attrs(
...
+        A function that is called with the original class object and all
+        fields: ``ft(cls: type, attribs: List[Attribute]) -> List[Attribute]``.
+        The hook is called right before the class gets transformed into an
+        attrs class.  The fields are already :class:`Attribute` instances.  You
+        can use this, e.g., to automatically add converters or validators to
+        fields.
```
Comment (hynek) [inline src/attr/_make.py:1217]: I'm gonna say we should shorten this to "what does it do and why it's interesting" and then link to extending:

```suggestion
        A function that is called with the original class object and all
        fields right before ``attrs`` finalizes the class.  You can use
        this, e.g., to automatically add converters or validators to
        fields based on their types.  See `transform-fields` for more details.
```

File: docs/extending.rst [hunk 16]
```diff
@@ -159,3 +159,107 @@ Here are some tips for effective use of metadata:
     ...     x = typed(int, default=1, init=False)
     >>> attr.fields(C).x.metadata[MY_TYPE_METADATA]
     <class 'int'>
+
```
Comment (hynek) [inline docs/extending.rst:162]: This is for a reference from the `attr.s` docstring below.

```suggestion


.. _transform-fields:
```

File: tests/test_hooks.py [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+import attr
+
+
+class TestTransformHook:
+    """Tests for `attrs(tranform_value_serializer=func)`"""
+
+    def test_hook_applied(self):
+        """
+        The transform hook must be applied to all attributes.  Types will not
+        be available here (they only work with type annotations).
```
Comment (hynek) [inline tests/test_hooks.py:13]: Wait, really? Why?

File: tests/test_hooks.py [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+
+        @attr.s(auto_attribs=True, field_transformer=hook)
+        class C:
+            x: int
+            y: str = attr.ib()
+
+        assert results == [("x", int), ("y", str)]
+
+    def test_hook_applied_modify_attrib(self):
+        """The transform hook must be able to modify attributes."""
```
Comment (hynek) [inline tests/test_hooks.py:47]:
```suggestion
        """
        The transform hook can modify attributes.
        """
```

File: tests/test_hooks.py [hunk 19]
```diff
@@ -0,0 +1,176 @@
+from datetime import datetime
+from typing import Dict, List
+
+import attr
+
+
+class TestTransformHook:
+    """Tests for `attrs(tranform_value_serializer=func)`"""
```
Comment (hynek) [inline tests/test_hooks.py:8]:
```suggestion
    """
    Tests for ``attr.s(transform_value_serializer=func)``
    """
```

File: tests/test_hooks.py [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+            x: int
+            y: str = attr.ib()
+
+        assert results == [("x", int), ("y", str)]
+
+    def test_hook_applied_modify_attrib(self):
+        """The transform hook must be able to modify attributes."""
+
+        def hook(cls, attribs):
+            return [a._assoc(converter=a.type) for a in attribs]
```
Comment (hynek) [inline tests/test_hooks.py:50]: We should probably make `_assoc` public then? 🤔 Or at least document it?

File: tests/test_hooks.py [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+        @attr.s(auto_attribs=True, field_transformer=hook)
+        class C:
+            x: int = attr.ib(converter=int)
+            y: float
+
+        c = C(x="3", y="3.14")
+        assert c == C(x=3, y=3.14)
+
+    def test_hook_remove_field(self):
+        """It must be possible to remove fields via the hook."""
```
Comment (hynek) [inline tests/test_hooks.py:61]:
```suggestion
        """
        It is possible to remove fields via the hook.
        """
```

File: tests/test_hooks.py [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+
+        @attr.s(auto_attribs=True, field_transformer=hook)
+        class C:
+            x: int
+            y: float
+
+        assert attr.asdict(C(2.7)) == {"y": 2.7}
+
+    def test_hook_add_field(self):
+        """It must be possible to remove fields via the hook."""
```
Comment (hynek) [inline tests/test_hooks.py:74]:
```suggestion
        """
        It is possible to add fields via the hook.
        """
```

File: tests/test_hooks.py [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+            return [a1, a2]
+
+        @attr.s(auto_attribs=True, field_transformer=hook)
+        class C:
+            x: int
+
+        assert attr.asdict(C(1, 2)) == {"x": 1, "new": 2}
+
+    def test_hook_with_inheritance(self):
+        """The hook must also receive all fields of base classes."""
```
Comment (hynek) [inline tests/test_hooks.py:88]:
```suggestion
        """
        The hook receives all fields from base classes.
        """
```

File: tests/test_hooks.py [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+        class Sub(Base):
+            y: int
+
+        assert attr.asdict(Sub(2)) == {"y": 2}
+
+
+class TestAsDictHook:
+    def test_asdict(self):
+        """
+        asdict() must call the hooks in attrs classes and in other
```
Comment (hynek) [inline tests/test_hooks.py:109]:
```suggestion
        asdict() calls the hooks in attrs classes and in other
```

File: tests/test_hooks.py [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+        assert result == {
+            "a": {"x": 1, "y": ["2020-07-01T00:00:00"]},
+            "b": [{"x": 2, "y": ["2020-07-02T00:00:00"]}],
+            "c": {"spam": {"x": 3, "y": ["2020-07-03T00:00:00"]}},
+            "d": {"eggs": "2020-07-04T00:00:00"},
+        }
+
+    def test_asdict_calls(self):
+        """
+        The correct instances and attribute names must be passed to the hook.
```
Comment (hynek) [inline tests/test_hooks.py:147]:
```suggestion
        The correct instances and attribute names are passed to the hook.
```

**Review by hynek — CHANGES_REQUESTED** [review]
A lot of nitpicking but thanks to the magic of suggestions, it should be quick to get thru!

_collapses on the floor_

File: docs/extending.rst [hunk 26]
```diff
@@ -159,3 +159,107 @@ Here are some tips for effective use of metadata:
...
+Automatic field transformation and modification
+------------------------------------------------
+
+Attrs allows you to automatically modify or transform the class' fields while the class is being created.
+You do this by passing a *field_transformer* hook to :func:`attr.define()` (and its friends).
+Its main purpose is to automatically add converters to attributes based on their type to aid the development of API clients and other typed data loaders.
+
+| This hook must have the signature:
+| :code:`your_hook(cls: type, fields: List[Attribute]) -> List[Attribute]`
```
[the remark this answers is not in the document]
Comment (sscherfke) [reply docs/extending.rst:172]: The pipes just enforce manual line breaks. Your solution seems a lot better. :)

File: src/attr/_funcs.py (hunk 12 above)
Comment (sscherfke) [reply src/attr/_funcs.py:38]: Narrator: It didn’t.

File: tests/test_hooks.py (hunk 17 above)
Comment (sscherfke) [reply tests/test_hooks.py:13]: This might no longer be true. I think that was the case with my original implementation. I will check this agian.

File: tests/test_hooks.py (hunk 20 above)
Comment (sscherfke) [reply tests/test_hooks.py:52]: Why not both? 😼

File: src/attr/_funcs.py (hunk 13 above)
Comment (sscherfke) [reply src/attr/_funcs.py:54]: Good point will change and document this.

File: tests/test_hooks.py (hunk 17 above)
Comment (sscherfke) [reply tests/test_hooks.py:15]: `_assoc()` is currently being used in these places:
```
src/attr/hooks.py
50:                attrib = attrib._assoc(converter=converter)

src/attr/_make.py
376:            a = a._assoc(inherited=True)
414:            a = a._assoc(inherited=True)
504:        own_attrs = [a._assoc(kw_only=True) for a in own_attrs]
505:        base_attrs = [a._assoc(kw_only=True) for a in base_attrs]
2326:    def _assoc(self, **changes):
```
I don’t think making it public would cause any harm, but you are the expert here. 😸

Comment (sscherfke) [reply tests/test_hooks.py]: I made it public and updated the docs. I also merged the additional docs from `api.rst` into the class’s docstring since the rendered docs looked and read really strange.

**Comment by sscherfke** [comment]
Ohai, this is the weekly ping. 🙊

**Review by hynek — APPROVED** [review]
We should change `_assoc` to `evolve`, not `assoc` but I’ll do it myself, to not delay this any further. Thanks and thank you for the patience, this is a big one!

**Comment by sscherfke** [comment]
Thanks for merging! ❤️ 

Next PR will arrive shortly 😬
